### ISSUE TITLE
Disruption of equities data :: pandas_datareader dependency on Yahoo and Google Finance API

### ISSUE THREAD

**Issue by rsvp** [rsvp/fecon235#7]
### Description of specific issue

We are expecting major disruption in getting data on 
*equities, mutual funds, and ETFs* via **pandas_datareader** 
due its dependency on API of both *Yahoo and Google Finance.*

- [/] Bug possibly, if no fix upstream
- [/] Enhancement, if fecon235 takes independent action

### Observed behavior

[Yahoo employee has confirmed that the free End-Of-Day data has been terminated, 2017](https://stackoverflow.com/questions/44057580/yahoo-finance-api-changes-2017)

[/r/algotrading on Google JSON](https://www.reddit.com/r/algotrading/comments/6ygdsg/google_cancelled_their_json_stock_data_feeds_last/)

### Steps which enhance behavior

1. Switch over to Quandl, using our lib/yi_quandl.py module
2. tiingo: https://api.tiingo.com/docs/general/overview
3. Alpha Vantage: https://www.alphavantage.co/documentation/
4. pandas_datareader issues: https://github.com/pydata/pandas-datareader/issues

### Why would the improvement be useful to most users?

The information from the equities markets is vital for financial economics.

Before releasing our own independent solution, we expect to 
make a pull request to the pandas_datareader repository.

***Please kindly propose alternative solutions.*** 

### Additional helpful details for bugs

- [/] Problem started recently, but not in older versions
- [/] Problem can be reliably reproduced

- **fecon235 version:** v5.17.0603
- **pandas version:** 0.19.2
- **pandas_datareader version:** 0.2.1

**Comment by rsvp** [comment]
## Reports of Google limiting data length to merely one year

- https://github.com/pydata/pandas-datareader/issues/395

- https://github.com/pydata/pandas-datareader/issues/394
    - Also: URL to retrieve data from Google has changed
       - Error messages from pandas_datareader 0.5

**Comment by paintdog** [comment]
What means:
> Our yi_stock module may appear to be working, but
>please tail() your dataframe to verify whether quote retrieval
>goes beyond 2017-09-05 [silent fail to get() current stock quotes].

Will there be no new quotes after 2017-09-05 or no data before 2017-09-05? I don't understand it.

I want to mention, that the new sources are not suitable for all users and do not cover all fields of interest :(

**Comment by rsvp** [comment]
@paintdog tail() retrieves the tail end of a dataframe. 
Our yi_stock module currently has a dependency on pandas_reader 
to get() *equities* data, but it may be re-route requests by other means 
(and to other vendors besides Google Finance).

All our sources for suitable for all users in the field of financial economics.

**Comment by rsvp** [comment]
[StackOverflow: DataReader google finance date not working](https://stackoverflow.com/questions/46259528/datareader-google-finance-date-not-working/) points to change in URL at Google as cause for data disruption.

But unfortunately, for the PR to modify said URL,  https://github.com/pydata/pandas-datareader/pull/402/commits/eac67a4b60c1f6b67f0a4323fa66faf49641b77b ,
the required Travis CI build has failed.

**Comment by rsvp** [comment]
Google Finance website currently displays a **Yellow Warning Banner** which reads as follows:

>  Google Finance is under renovation. As a part of this process, the Portfolios feature won't be available after ***mid-November 2017***.

Today is 27 September, so we are talking roughly ***seven weeks under renovation***. 
So any fixes now may go to waste after such renovation: 
- [  ] wait and see the renovated source API for upstream repo.

Noteworthy is the redirect from the *www* to *finance* subdomain for Google.

**Comment by rsvp** [comment]
*Upstream fixes not passing the tests in Travis CI build* has been **resolved**: https://github.com/pydata/pandas-datareader/pull/404

This is especially pertains to:
https://github.com/pydata/pandas-datareader/issues/394
https://github.com/pydata/pandas-datareader/issues/395
https://github.com/pydata/pandas-datareader/pull/402

**Comment by paintdog** [comment]
What does 
> pytest.skip("Google quote api is offline as of Oct 1, 2017")
means? At the moment I can still retrieve data from Google!

**Comment by rsvp** [comment]
@paintdog That line is used to avoid testing designated parts of the 
Travis CI build upstream at pandas-datareader. In the interim, 
suppressing the bug catchers is not a good idea...

As you have observed the Google API still operational, 
so the use of pytest.skip for the Google Finance test fix is indeed
*too definitive* in the sense that Google may reconsider its API changes
during the reconstruction period. Tests could use pytest.xfail or even more
explicitly `@pytest.mark.xfail()` with condition and reason attached.
See https://docs.pytest.org/en/latest/skipping.html

That way upstream, we can later run: `pytest --runxfail`
to see if Google has reverted without much hassle 
and just remove the decorators, or rewrite tests 
to conform to the new API.
(Historically, Google has reversed its pronouncements on quote
retrieval via Sheets, so there's hope if there's enough public outcry :-).

**Comment by rsvp** [comment]
## Alternative: Morningstar

- [ ] MORNINGSTAR API

Some work has begun: https://github.com/pydata/pandas-datareader/pull/411
which includes quotes and data from ***financial statement***.
But the Travis CI build has failed.

At fecon235, we may integrate partial code directly from:
https://gist.github.com/hahnicity/45323026693cdde6a116 
which is far more informative than anything from the 
origin https://www.moringstar.com

**Comment by rsvp** [comment]
- [ ] Test pandas-datareader: Change from *skip* to *mark.xfail* in test_google

See  https://github.com/pydata/pandas-datareader/pull/413

**Comment by rsvp** [comment]
## Alternative: Tiingo

- [ ] TIIANGO API

Tiingo has a REST and Real-Time Data API includes support for
Stock Market Ticker Closing Prices + Metadata.

At fecon235, we may integrate partial code directly from:
https://github.com/hydrosquall/tiingo-python
which outputs CSV or JSON, so conversion to DataFrame
format is the next step. Thanks to @hydrosquall

**Comment by rsvp** [comment]
## Alternative: Alpha Vantage

- [ ] Alpha Vantage API 

Documentation https://www.alphavantage.co/documentation 
Requires user to get free API key

@RomelTorres [commented](https://github.com/pydata/pandas-datareader/issues/389#issuecomment-339352735):

> I don't think AlphaVantage itself will release a python wrapper for its API call.

Hence his Python module to get stock data and cryptocurrencies: 
https://github.com/RomelTorres/alpha_vantage

We would like to invite more discussion comparing the data vendors 
at Gitter https://gitter.im/rsvp/fecon235

**Comment by rsvp** [comment]
## Alternative, not supported: Bloomberg API

Documentation  https://www.bloomberg.com/professional/support/api-library/
Requires user to have Bloomberg Terminal access, 
which will *cost $$$ (thus does not share core fecon235 objectives)*.

For Python pandas interface, we refer to the repo by @matthewgilbert 
https://github.com/matthewgilbert/pdblp

**Comment by rsvp** [comment]
## pandas-datareader RemoteDataError

For 0.5.0 fails reported for both `get_quote_yahoo` and `get_quote_google`,
see https://github.com/pydata/pandas-datareader/issues/418

## Alternative: Barchart API

Documentation:  https://www.barchart.com/ondemand/api
Requires registration: https://www.barchart.com/ondemand/free-market-data-api
Features stock, **futures, FX, and cryptocurrencies** 
(presumably the CME Bitcoin contract upon regulatory approval).
*Free* histories up to 2 years, where End-of-day available for: 

- Equities: AMEX, LSE, NASDAQ, NYSE
- Futures: CBOT, CME, COMEX, NYMEX
- Forex: Global Currencies

- [  ] Barchart API, CSV -> Python pandas DataFrame (cc: @liuyigh)

**Comment by paintdog** [comment]
> For 0.5.0, failures reported for both get_quote_yahoo and get_quote_google,
> see pydata/pandas-datareader#418 as of 2017-11-07.

This is too bad - no hope that Google will deliver data in the soon future again???

**Comment by rsvp** [comment]
**Re: [Thanks for all the fish](http://www.financial-hacker.com/bye-yahoo-and-thank-you-for-the-fish/)**

- **Stooq** (based in Poland), on first impression, does seem suitable for fecon235.

> [F]or ETFs they [Quandl] want 50 bucks per month.

- [Caltech Quantitative Finance Group](http://quant.caltech.edu/historical-stock-data.html) provides notes on data vendors, esp. those to avoid.

- [ ] investigate QuantQuote, https://quantquote.com/historical-stock-data

> Scraping by URL: The ***cookie field*** is not for preventing robots, but for disabling all API access by software. You can indeed step around the cookie method with some tricks, even automated, but that’s not really a professional solution. // That “crumb” in your URL is a code connected to a cookie on your PC. No cookie, no data.

> Alpha Vantage does not adjust for splits and dividends. Their reply: "We have just beta-launched the adjusted close price feature for our daily prices API."

 - Alpha Vantage has a dedicated JSON API service for split/dividend-adjusted EOD historical data. For example: https://www.alphavantage.co/query?function=TIME_SERIES_DAILY_ADJUSTED&symbol=MSFT&outputsize=full&apikey=demo

> Jenn: While Alpha Vantage has great data and a well-documented API, it’s pretty sketchy that they don’t describe their business model or have an “About Us” page that gives a company address and information about the actual human beings who work there. I’d love to use it, but it’s hard to trust it at the moment.

**Comment by rsvp** [comment]
## Yahoo cookie / crumb : Python solutions

Crumb is just part of the cookie, and here are some Python scrape solution:

- @lucas03 provided a [gist](https://stackoverflow.com/posts/44445027/revisions) at StackOverflow
- @bradlucas blog posted [code](http://blog.bradlucas.com/posts/2017-06-03-yahoo-finance-quote-download-python/)
- @sjev summarized what works in a [Jupyter notebook](https://github.com/sjev/trading-with-python/blob/master/scratch/get_yahoo_data.ipynb) which is ideal for **testing revised URL changes at Yahoo**

**Comment by rsvp** [comment]
### Data sources

@wilsonfreitas provides an extensive listing:
https://github.com/wilsonfreitas/awesome-quant#data-sources

Any preferences therein which are reliable?

**Comment by rsvp** [comment]
## Alternative: IEX API

- No key registration.
- Free data for redistribution.
- Supports websockets, i.e. pushes data.

Main page: https://iextrading.com/developer
API: https://iextrading.com/developer/docs/#getting-started

- [ ] GitHub: https://github.com/iexg/IEX-API

Thanks to @iexg and @lockefox

**Comment by rsvp** [comment]
#### Some *interim* Google functionality: November 2017

- Get a CSV file: sample URL from @VicTangg, https://finance.google.com/finance/historical?q=SPY&startdate=2017/01/01&enddate=2017/05/22&output=csv

- [pandas_datareader snippet](https://github.com/pydata/pandas-datareader/issues/418#issuecomment-346433713) from @paintdog :

```python
import pandas_datareader.data as web
import datetime

start = datetime.datetime(2017, 1, 1)
end = datetime.date.today()

google = False

if google:
    f = web.DataReader("ETR:SIE", 'google', start, end)
else:
    f = web.DataReader("SIE.DE", 'yahoo', start, end)

print(f.Close)
```

**Comment by paintdog** [comment]
I hope that the info from @VicTangg will be used to repair pandas datareader. It seems that Google is still delivering data in an acceptable quality.

**Comment by bsolomon1124** [comment]
A new issue has surfaced in the last few days regarding google pulls with `pandas-datareader`.  (See [424](https://github.com/pydata/pandas-datareader/issues/424).)

However, the API itself still seems intact:

```python
import datetime
import requests
from io import StringIO
# This is just a wrapper importing the compatible version of
#     urllib's urlencode--see pandas docs
from pandas.io.common import urlencode
import pandas as pd

BASE = 'http://finance.google.com/finance/historical'


# There seems to be confusion over whether the date api has changed.
# https://github.com/pydata/pandas-datareader/pull/425
# Both formats seem to work, but I'll use the "newer" one here to be safe
def get_params(symbol, start, end):
    params = {
        'q': symbol,
        'startdate': start.strftime('%Y/%m/%d'),
        'enddate': end.strftime('%Y/%m/%d'),
        'output': "csv"
    }
    return params


def build_url(symbol, start, end):
    params = get_params(symbol, start, end)
    return BASE + '?' + urlencode(params)


start = datetime.datetime(2010, 1, 1)
end = datetime.datetime.today()
sym = 'SPY'
url = build_url(sym, start, end)

data = requests.get(url).text
data = pd.read_csv(StringIO(data), index_col='Date', parse_dates=True)

print(data.head())
#               Open    High     Low   Close     Volume
# Date
# 2017-11-30  263.76  266.05  263.67  265.01  127894389
# 2017-11-29  263.02  263.63  262.20  262.71   77512102
# 2017-11-28  260.76  262.90  260.66  262.87   98971719
# 2017-11-27  260.41  260.75  260.00  260.23   52274922
# 2017-11-24  260.32  260.48  260.16  260.36   27856514
```

**Comment by rsvp** [comment]
@bsolomon1124 hi Brad, I would be inclined to think that 
**Google intends UTF-8, not bytes, over the wire**, so thanks 
for your modification from `requests.get(url).content`
to `requests.get(url).text`

In regarding to your proposals for **Google URL** in 
https://github.com/pydata/pandas-datareader/pull/347 and https://github.com/pydata/pandas-datareader/pull/425
`%2F` within the URL is just the longer machine version 
of `-` which is human readable, e.g.:
http://finance.google.com/finance/historical?q=SPY&startdate=2010%2F01%2F01&enddate=2017%2F12%2F01&output=csv 

Brad, your fixes are very much appreciated. *Thank you!*

**Comment by bsolomon1124** [comment]
@rsvp I'm still not sure if bytes v. string is the issue here, though, the more I look into it.  Reading in bytes is [explicitly covered/addressed](https://github.com/pydata/pandas-datareader/blob/master/pandas_datareader/base.py#L86).  Instead, I'm starting to think it's just that the GitHub code is [not reflected in PyPI](https://github.com/pydata/pandas-datareader/issues/426).

I.e.:

```python
import datetime

from pandas.compat import StringIO, bytes_to_str
from pandas.io.common import urlencode
import requests


BASE = 'http://finance.google.com/finance/historical'


def _get_params(symbol, start, end):
    params = {
        'q': symbol,
        'startdate': start.strftime('%Y/%m/%d'),
        'enddate': end.strftime('%Y/%m/%d'),
        'output': "csv"
    }
    return params


def build_url(symbol, start, end, form='new'):
    params = _get_params(symbol, start, end)
    return BASE + '?' + urlencode(params)


sym = 'AAPL'
start = date(2010, 1, 1)
end = date.today()

url = build_url(sym, start, end)
# http://finance.google.com/finance/historical?q=AAPL&startdate=Jan+01%2C+2010&enddate=Dec+05%2C+2017&output=csv


session = requests.Session()
byts = session.get(url).content
out = StringIO()
out.write(bytes_to_str(byts))
out.seek(0)
data = pd.read_csv(out, index_col=0, parse_dates=True).sort_index()

data 
              Open    High     Low   Close     Volume
Date                                                 
2010-01-04   30.49   30.64   30.34   30.57  123432050
2010-01-05   30.66   30.80   30.46   30.63  150476004
2010-01-06   30.63   30.75   30.11   30.14  138039594
2010-01-07   30.25   30.29   29.86   30.08  119282324
2010-01-08   30.04   30.29   29.87   30.28  111969081
2010-01-11   30.40   30.43   29.78   30.02  115557365
2010-01-12   29.88   29.97   29.49   29.67  148614774
2010-01-13   29.70   30.13   29.16   30.09  151472335
2010-01-14   30.02   30.07   29.86   29.92  108288411
2010-01-15   30.13   30.23   29.41   29.42  148584065
2010-01-19   29.76   30.74   29.61   30.72  182501620
2010-01-20   30.70   30.79   29.93   30.25  153037892
           ...     ...     ...     ...        ...
2017-11-16  171.18  171.87  170.30  171.10   23637484
2017-11-17  171.04  171.39  169.64  170.15   21899544
2017-11-20  170.29  170.56  169.56  169.98   16262447
2017-11-21  170.78  173.70  170.78  173.14   25131295
2017-11-22  173.36  175.00  173.05  174.96   25588925
2017-11-24  175.10  175.50  174.65  174.97   14026673
2017-11-27  175.05  175.08  173.34  174.09   20716802
2017-11-28  174.30  174.87  171.86  173.07   26428802
2017-11-29  172.63  172.92  167.16  169.48   41666364
2017-11-30  170.43  172.14  168.44  171.85   41527218
2017-12-01  169.95  171.67  168.50  171.05   39759288
2017-12-04  172.48  172.62  169.63  169.80   32542385
```

**Comment by rsvp** [comment]
@bsolomon1124 hi Brad, interesting detective work there with Py**PI**.

The question of code replication would then extend to those who have 
installed pandas_datareader via **Anaconda**'s `conda install` command. 
By further extension, it would apply to those using Anaconda's Docker image 
to create *containers* where such an auto-install is scripted.

**Comment by bsolomon1124** [comment]
@rsvp Not just `conda`, but `pip` also.  Take a look in the [tar](https://pypi.python.org/packages/9b/3b/5fae1f4ba5889468fefbadcefda0389f60f2780eead3a1166415bda4509c/pandas-datareader-0.5.0.tar.gz#md5=77cc22b08b5cac46eb974ac6687d4555) from PyPI and you'll notice the old link in place.  (`pandas-datareader/google/daily/_GoogleDailyReader.url`)

**Comment by rsvp** [comment]
#### Some fixes on forthcoming pandas_datareader 0.6.0 

Thanks to @davidastephens 

#### Yahoo: cryptocurrency quotes

When people are taking out mortgages on their homes to make bets...

- https://finance.yahoo.com/cryptocurrencies lists a hundred out of 1300 (and growing).

- https://finance.yahoo.com/cryptocurrencies/heatmap Heatmap of capitalizations percentage gain/loss.

**Comment by rsvp** [comment]
## UPGRADE to development version of pandas_datareader 

Is there a way to mitigate interim disruptions with Yahoo
and Google, and get the latest documentation, without waiting
around for a major release? Often a fix simply involves 
revising an URL address, but the fix arrives downstream 
months later.

### Anaconda distribution

It is possible to `conda` install the HEAD of the master branch: 
https://conda.io/docs/commands/build/conda-develop.html 
Use `conda develop /path/to/pandas_datareader` 
Install a Python package in 'development mode'.
This works by creating a conda.pth file in site-packages. 
Even `--uninstall` is provided to back out of development.
Thanks to [@ralexx](https://github.com/pydata/pandas-datareader/issues/443#issuecomment-358035786)

@bashtage notes:
> Conda-forge can be triggered by users though, and since this package doesn't have any binary requirements, PyPi is as good as anything else.

### Direct from source

```shell
#  To get master HEAD:
git clone https://github.com/pydata/pandas-datareader
cd pandas-datareader
python setup.py install
```

It is an *open question* whether setup.py will *properly* 
install with respect to both the system's and Anaconda's 
versions of Python.

**Comment by rsvp** [comment]
### pandas_reader Deprecations

As of 2018-01-18, leading up to their 0.6.0 release, contributor @bashtage 
has deprecated:

- [Yahoo finance] (https://github.com/pydata/pandas-datareader/issues/418#issuecomment-358699604) and introduced new label [yahoo-finance](https://github.com/pydata/pandas-datareader/labels/yahoo-finance)
- [Google finance](https://github.com/pydata/pandas-datareader/issues/422#issuecomment-358699683) and introduced new label [google-finance](https://github.com/pydata/pandas-datareader/labels/google-finance)

Fallback vendor(s) for reliable equities data has not been clarified 
at https://github.com/pydata/pandas-datareader thus fecon235 
may have to develop independently.

**Comment by rsvp** [comment]
## Alternative: Robinhood

- [ ] Robinhood API 

Documentation https://github.com/sanko/Robinhood
No API key required. Thanks to @sanko

Snippet for current quotes:
key https://api.robinhood.com/quotes/?symbols=MSFT,FB,TSLA

Snippet for historical data:
https://api.robinhood.com/quotes/historicals/?symbols=MSFT,FB,TSLA&interval=day

**Comment by rsvp** [comment]
## REFERENCES for new Alternatives

- [pandas-datareader, "see-also" docs ](https://github.com/pydata/pandas-datareader/blob/master/docs/source/see-also.rst)
    - [Label: new-data-source](https://github.com/pydata/pandas-datareader/labels/new-data-source)

- [awesome-quant, section on data sources](https://github.com/wilsonfreitas/awesome-quant#data-sources)

**Comment by rsvp** [comment]
## Alternative: Interactive Brokers

Official: https://interactivebrokers.github.io

The [official Interactive Brokers Python API](https://github.com/InteractiveBrokers/tws-api-public) has a few design choices which make it run slowly. Specifically: excessive debug logging, and an overly cautious lock on the socket connection. [ibapi-grease](https://github.com/quantrocket-llc/ibapi-grease) provides monkey patches that eliminate these bottlenecks by turning off the logging and removing the locks. -- @quantrocket-llc 

IB's Zipline only supports backtesting, while IB's Catalyst supports backtesting and live trading. Zipline and Catalyst support a separate data ingestion step which downloads the data once:

http://www.zipline.io/bundles.html
https://github.com/quantopian/zipline/blob/master/zipline/data/
https://github.com/enigmampc/catalyst/tree/master/catalyst/data

Above, thanks to [@westurner](https://github.com/pydata/pandas-datareader/issues/454)

**Comment by rsvp** [comment]
## [pandas_datareader v0.6.0 Release](https://github.com/pydata/pandas-datareader/releases/tag/v0.6.0)

@bashtage released this 3 hours ago -- tremendous work!

**Warning:** Yahoo!, Google Options, Google Quotes and EDGAR have been immediately deprecated.

But Google finance is still functioning for historical price data, although there are frequent reports of failures. *Google failure is frequently encountered when bulk downloading historical price data.*

Highlights include:

- Immediate deprecation of Yahoo!, Google Options and Quotes and EDGAR. The end points behind these APIs have radically changed and the existing readers require complete rewrites. In the case of most Yahoo! data the endpoints have been removed. PDR would like to restore these features, and pull requests are welcome.
 
- A new connector for Tiingo was introduced. Tiingo provides historical end-of-day data for a large set of equities, ETFs and mutual funds. Free registration is required to get an API key.

- A new connector for Robinhood was introduced. This provides up to 1 year of historical end-of-day data. It also provides near real-time quotes.

- A new connector for Morningstar Open, High, Low, Close and Volume was introduced.

- A new connector for IEX daily price data was introduced. The majority of the IEX API was introduced.

- A new data connector for stock index data provided by Stooq was introduced.

- A new data connector for data provided by the Bank of Canada was introduced.

**Comment by rsvp** [comment]
2018-03-20 @jfunction in https://github.com/pydata/pandas-datareader/issues/502#issuecomment-374587780
***reports Google Finance has removed historical data***.
In trying: http://finance.google.com/finance/historical?q=GOOG 
we got the following message on that webpage:

> We're sorry... but your computer or network may be sending automated queries. To protect our users, we can't process your request right now. See Google Help for more information.

The Google Help page: https://support.google.com/websearch/answer/86640 
in its relevant portion reads as follows:

#### "Unusual traffic from your computer network"

```
If devices on your network seem to be sending automated traffic to Google, 
you might see "Our systems have detected unusual traffic from your computer network."

What Google considers automated traffic:
- Sending searches from a robot, computer program, automated service, or search scraper
- Using software that sends searches to Google to see how a website or webpage ranks on Google
```

In other words, it *seems that Google Finance is now only intended for human eyes*, 
not automated scripts. Our guess: for their advertising revenue.

**Comment by rsvp** [comment]
## CHANGELOG 2018-06-23  (tag: v6.18.0623)

**Major version change for fecon235 from v5 to v6**
signaling our integration with **fecon236**
which was spun-off from our source code.

**Henceforth, fecon235 becomes a repository solely of Jupyter notebooks.**
The old Python source code at fecon235 will remain for archival purposes,
while new code development shifts over to fecon236.
Please see https://git.io/econ for details.

Revise docs/fecon235-00-README.ipynb to introduce fecon236.
Function names have been retained, but under fecon236
the call routing is expected to be more explicit than casual,
i.e. modules names are more significant.
Please see https://git.io/fecon-intro

# NOTICE of MOVE

This issue and its remedies has moved to: https://github.com/MathSci/fecon236/issues/2

**Your review and feedback there would be greatly appreciated. Thank you!**

**Comment by rsvp** [comment]
### Possibility of using QUANDL is likely nill for freely accessible data

... since *Nasdaq acquired Quandl on or about December 4, 2018*, 
and appears to have turned proprietary.

For full details, see follow-up on **"Disruption of Quandl data, esp. futures"**
https://github.com/MathSci/fecon236/issues/7
